packages/data/src/components/with-dispatch/index.js

```diff
@@ -30,46 +24,10 @@ const withDispatch = ( mapDispatchToProps ) => createHigherOrderComponent(
 	compose( [
 		pure,
 		( WrappedComponent ) => {
-			class ComponentWithDispatch extends Component {
-				constructor( props ) {
-					super( ...arguments );
-
-					this.proxyProps = {};
-					this.setProxyProps( props );
-				}
-
-				componentDidUpdate() {
-					this.setProxyProps( this.props );
-				}
-
-				proxyDispatch( propName, ...args ) {
-					// Original dispatcher is a pre-bound (dispatching) action creator.
-					mapDispatchToProps( this.props.registry.dispatch, this.props.ownProps )[ propName ]( ...args );
-				}
-
-				setProxyProps( props ) {
-					// Assign as instance property so that in subsequent render
-					// reconciliation, the prop values are referentially equal.
-					// Importantly, note that while `mapDispatchToProps` is
-					// called, it is done only to determine the keys for which
-					// proxy functions should be created. The actual registry
-					// dispatch does not occur until the function is called.
-					const propsToDispatchers = mapDispatchToProps( this.props.registry.dispatch, props.ownProps );
-					this.proxyProps = mapValues( propsToDispatchers, ( dispatcher, propName ) => {
-						// Prebind with prop name so we have reference to the original
-						// dispatcher to invoke. Track between re-renders to avoid
-						// creating new function references every render.
-						if ( this.proxyProps.hasOwnProperty( propName ) ) {
-							return this.proxyProps[ propName ];
-						}
-
-						return this.proxyDispatch.bind( this, propName );
-					} );
-				}
+			function ComponentWithDispatch( { registry, ownProps } ) {
+				const mergeProps = mapDispatchToProps( registry.dispatch, ownProps );
```

Comment:
Any concerns about the new props we keep creating with this? (new instances of functions on each call?)

Reply:
> Any concerns about the new props we keep creating with this? (new instances of functions on each call?)

Yep, it's a trade-off between that and iterating to create `proxyProps`. I think there's a middle-ground option where we use `static getDerivedStateFromProps` and still defer to a proxied implementation. The proxying in general adds a bit of complexity, and in the end I'm not sure if it's saving us much in the end.

Reply:
(FYI this is also expanded upon in the original comment)